pallets/dispenser/src/lib.rs

```diff
@@ -0,0 +1,187 @@
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		/// The loose coupling to a vesting schedule implementation.
+		type VestingSchedule: VestingSchedule<Self::AccountId, Moment = BlockNumberFor<Self>>;
+		/// The period of time that the dispensed funds are in a vesting schedule. The schedule
+		/// starts after the lock period.
+		#[pallet::constant]
+		type VestPeriod: Get<BlockNumberFor<Self>>;
+		/// The Ed25519 Verifier Public Key to verify the signature of the credentials.
+		#[pallet::constant]
+		type VerifierPublicKey: Get<[u8; 32]>;
+		/// A type representing the weights required by the dispatchables of this pallet.
+		type WeightInfo: crate::weights::WeightInfo;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	pub type DispenseAmount<T> = StorageValue<_, BalanceOf<T>, ValueQuery, <T as Config>::InitialDispenseAmount>;
+
+	#[pallet::storage]
+	pub type Dispensed<T> = StorageMap<_, Blake2_128Concat, Did, ()>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		Dispensed { dispensed_to_did: Did, dispensed_to: T::AccountId, amount: BalanceOf<T> },
+		DispenseAmountChanged(BalanceOf<T>),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The dispenser has already dispensed to the did.
+		DispensedAlreadyToDid,
+		/// The dispenser account does not have any funds to distribute.
+		DispenserDepleted,
+		/// The dispense amount is too low. It must be greater than the free dispense amount.
+		DispenseAmountTooLow,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::feeless_if( | origin: &OriginFor<T>, jwt: &UntrustedToken | -> bool {
+            if let Ok((_, did, _)) = T::InvestorOrigin::ensure_origin(origin.clone(), jwt, T::VerifierPublicKey::get()) {
+                return Dispensed::<T>::get(did).is_none()
+            } else {
+                return false
+            }
+        })]
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::dispense())]
+		pub fn dispense(origin: OriginFor<T>, jwt: UntrustedToken) -> DispatchResultWithPostInfo {
+			let (who, did, _investor_type) =
+				T::InvestorOrigin::ensure_origin(origin, &jwt, T::VerifierPublicKey::get())?;
+			ensure!(Dispensed::<T>::get(&did).is_none(), Error::<T>::DispensedAlreadyToDid);
+
+			let amount = DispenseAmount::<T>::get();
+			ensure!(CurrencyOf::<T>::free_balance(&Self::dispense_account()) >= amount, Error::<T>::DispenserDepleted);
+
+			let current_block = <frame_system::Pallet<T>>::block_number();
+			let length_as_balance = T::BlockNumberToBalance::convert(T::VestPeriod::get());
+			let locked_amount = amount.saturating_sub(T::FreeDispenseAmount::get());
+			let per_block = locked_amount / length_as_balance.max(sp_runtime::traits::One::one());
+
+			T::VestingSchedule::can_add_vesting_schedule(
```

Comment:
Its described in the vesting pallet implementation that it should always be called before the add_vesting_schedule. But it might be that its older code that was written before the transactionality of extrinsics, where the add_vesting_schedule should succeed after the transfer of tokens. I could probably remove it if you want?

Reply:
My only worry is that we already encountered the possibility of executing logic in a non-transactional way, and that was running functions that touched storage inside on_initialize.

I think if we are sure this will always run inside an extrinsic which has transacitonal by default, then we can be pretty sure it will roll back on error and remove the can_add_vesting_schedule. But I cannot tell you 100% it will be transactional.

I leave it up to you if you want to remove it

Reply:
Lets keep it for now. Might do some double checks but its fine. Pretty sure this logic will mainly stay in an extrinsic as it is part of an extrinsic.